### PR TITLE
build.d: Fix host compiler detection for LDC

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
The output of ldc2 --version contains the compiler used to build it (e.g. dmd) which results in HOST_DMD_KIND being set to "dmd".

Both DMD and LDC start the output of --version with their unqualified name (tested on Windows/Ubuntu)

### CODE REVIEW

**Review by wilzbach — CHANGES_REQUESTED** [review]
This is not necessarily always the case. Why not do a `find` with multiple needles? You'll get the index of the match back if there's one.

**Comment by MoonlightSentinel** [comment]
> This is not necessarily always the case. Why not do a `find` with multiple needles? You'll get the index of the match back if there's one.

I will look into that. 

On a side note, does the first line of `$(HOST_DMD_RUN) --version` always contain the currently used compiler?

**Comment by wilzbach** [comment]
> On a side note, does the first line of $(HOST_DMD_RUN) --version always contain the currently used compiler?

Yeah. I think it currently does, but I am not sure it's a good idea to depend on this as it might lead to similar breakage to the one that you just observed.

**Comment by MoonlightSentinel** [comment]
Updated based on your suggestion to use canFind with multiple needles.

File: src/build.d [hunk 1]
```diff
@@ -679,15 +679,17 @@ void processEnvironment()
 {
     const os = env["OS"];
 
-    auto hostDMDVersion = [env["HOST_DMD_RUN"], "--version"].execute.output;
-    if (hostDMDVersion.canFind("DMD"))
+    const hostDMDVersion = [env["HOST_DMD_RUN"], "--version"].execute.output;
+    const kindIdx = hostDMDVersion.canFind("DMD", "LDC", "GDC", "gdmd", "gdc");
+
+    enforce(kindIdx, "Invalid Host DMD found: " ~ hostDMDVersion);
```
Comment (marler8997) [inline src/build.d]: Having to keep the `kindIdx` in sync with the argument list above doesn't seem ideal.  What about something like this?
```D
    const hostDMDVersion = [env["HOST_DMD_RUN"], "--version"].execute.output;
    for (string versionRemaining = hostDMDVersion; ; versionRemaining.popFront) {
        enforce(!versionRemaining.empty, "Invalid Host DMD found: " ~ hostDMDVersion);
        if (versionRemaining.startsWith("DMD")) {
            env["HOST_DMD_KIND"] = "dmd";
            break;
        } else if (versionRemaining.startsWith("LDC")) {
            env["HOST_DMD_KIND"] = "ldc";
            break;
        } else if (versionRemaining.startsWith("GDC") ||
            versionRemaining.startsWith("gmd") ||
            versionRemaining.startsWith("gdb")) {
            env["HOST_DMD_KIND"] = "gdc";
            break;
        }
    }
```

Comment (MoonlightSentinel) [reply src/build.d]: I removed the magic numbers so we can add/remove other identifiers (e.g. "dmd") without breaking the evaluation of kindIdx.
If you would still prefer your proposed solution i can use that approach as well.

Comment (marler8997) [reply src/build.d]: Clever.  I like it better than the "magic numbers" solution.

I think the solution I proposed is a bit less brittle as you don't have to keep the "ordering" in sync and if you add a new case, you don't have to modify the other cases, i.e. with your solution if you added another kind before `DMD` or `LDC` then you would need to modify the `if` conditions for `DMD` and `LDC`. I'll defer to @wilzbach on the final decision.

Comment (MoonlightSentinel) [reply src/build.d]: True. Im fine with either solution as long as it works

Comment (wilzbach) [reply src/build.d]: @marler8997's solution is more elegant, but this is a build script that is supposed to be read by humans and there are only three compiler. This won't change in the foreseeable future, so I think we are fine with @MoonlightSentinel's solution. Thanks!

**Review by wilzbach — APPROVED** [review]

File: src/build.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/build.d]: Thank you @wilzbach and @marler8997.

**Pull request merged**